# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [ ]:
import pandas as pd

In [ ]:
# Your code here
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
cust_data_df = pd.read_csv(url)
cust_data_df

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [ ]:
#Challenge I Part I

num_rows, num_cols = cust_data_df.shape
print("Number of rows", num_rows)
print("Number of cols", num_cols)

Number of rows 4008
Number of cols 11


In [ ]:
#Challenge I Part II

cust_data_df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [ ]:
'''
Should be defined as an int or float:
  CLV, number of open complaints


In [ ]:
#Challenge I Part III

print(cust_data_df.describe())


             Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678


'''
Standard deviation = SD
Observations for data with int/float

---


**Average client profile**
The majority of the 1071 clients in this csv file (not accounting for deficiencies):
  come from Oregon
  are female
  have a bachelors degree
  the CLV is approximately $445811.34
  have a personal-auto insurance policy
  own 4 door cars


---
**Observations**

Income observation shows that the average income is around $39296 whilst the median is $36234. This means that there are outliers that push the mean up
The SD suggests that the data is erroneous as it would be unlikely someone with $0 would have a vehicle with insurance. The min and max suggest a huge range of income

Monthly premium auto observation suggests that the mean monthly premium is $193 whilst the median is lower at $83, this means a small number of individuals are paying a higher premium
The SD is high at $1,601 and the max at $99960 seems to suggest some individuals pay huge premiums, thus outliers.

Total Claim Amount suggests that the mean claim is around $405 whilst the median is $354.7 suggesting a moderate positive skew. The range whilst not as vast as the Income disparities suggest amounts from $0.38 to $2893, taking into account the 75th percentile claim amount is $532.8, suggesting several higher claims driving up the average



---

**Improvements**
-Investigate individuals with 0 income as potential errors
-Handle outliers in Monthly Premium auto
-Possible investigation into risk assessment via age if possible to procure data
  Maybe possible to segment customers based on claim amounts to help adjust premiums



In [ ]:
#Challenge I Part IV
print(cust_data_df.describe(include="all"))

       Customer      ST GENDER Education Customer Lifetime Value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   
mean        NaN     NaN    NaN       NaN                     NaN   
std         NaN     NaN    NaN       NaN                     NaN   
min         NaN     NaN    NaN       NaN                     NaN   
25%         NaN     NaN    NaN       NaN                     NaN   
50%         NaN     NaN    NaN       NaN                     NaN   
75%         NaN     NaN    NaN       NaN                     NaN   
max         NaN     NaN    NaN       NaN                     NaN   

              Income  Monthly Premium Auto Number of Open Complaints  \
count    1071.000000           1071.000000                      1071   
unique           NaN                   

''
Standard deviation = SD
Observations for all data incl strings

---
**Observations**

Customer - 1071 unique values which equals to number of entries. Showing that data is correct.

ST - Oregon appears most commonly, 320 times, suggesting a risk-taking behaviour in the state.

Education - Bachelor appears 324 times, suggesting that the client base is well educated.

Number of complaints - "1/0/00" appears as erroneous data and should be investigated or cleaned

Policy type - Personal auto appears the most at 780 times, suggesting that the sample size is suspect or may be less represented.

---

**Improvements**
- The Customer data list has a diverse mix in education, income and premium payments. This could be useful for customer segmentation and suggesting different pricing packages for different profiles within the customer base

-Outliers need investigation in several columns



In [1]:
#Challenge I Part V
#Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

import pandas as pd

# Load the data
df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')

# Identify categorical columns
categorical_cols = df.select_dtypes(include=['object']).columns

# Compute summary statistics for categorical columns
summary_stats = df[categorical_cols].describe()

# Display the summary statistics
print("Summary Statistics for Categorical Columns:")
print(summary_stats)

# Conclusions based on summary statistics
for col in categorical_cols:
    print(f"\nConclusions for '{col}':")
    print(f"- Number of unique values: {summary_stats[col]['unique']}")
    print(f"- Most frequent value (top): {summary_stats[col]['top']} (occurs {summary_stats[col]['freq']} times)")
    print(f"- Frequency distribution:\n{df[col].value_counts()}\n")



Summary Statistics for Categorical Columns:
       Customer      ST GENDER Education Customer Lifetime Value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   

       Number of Open Complaints    Policy Type  Vehicle Class  
count                       1071           1071           1071  
unique                         6              3              6  
top                       1/0/00  Personal Auto  Four-Door Car  
freq                         830            780            576  

Conclusions for 'Customer':
- Number of unique values: 1071
- Most frequent value (top): RB50392 (occurs 1 times)
- Frequency distribution:
Customer
RB50392    1
HJ15383    1
AO74776    1
HQ82233    1
OL72737    1
          ..
RO26085    1
ES57969    1
JK55587    1
RN97635    1
MY31220   

Conclusions for 'Customer':
- Number of unique values: 1071
- Most frequent value (top): RB50392 (occurs 1 times)
- Frequency distribution:
Customer
RB50392    1
HJ15383    1
AO74776    1
HQ82233    1
OL72737    1
          ..
RO26085    1
ES57969    1
JK55587    1
RN97635    1
MY31220    1
Name: count, Length: 1071, dtype: int64


Conclusions for 'ST':
- Number of unique values: 8
- Most frequent value (top): Oregon (occurs 320 times)
- Frequency distribution:
ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64


Conclusions for 'GENDER':
- Number of unique values: 5
- Most frequent value (top): F (occurs 457 times)
- Frequency distribution:
GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64


Conclusions for 'Education':
- Number of unique values: 6
- Most frequent value (top): Bachelor (occurs 324 times)
- Frequency distribution:
Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64


Conclusions for 'Customer Lifetime Value':
- Number of unique values: 1027
- Most frequent value (top): 445811.34% (occurs 4 times)
- Frequency distribution:
Customer Lifetime Value
445811.34%     4
251459.20%     4
272535.64%     3
578018.22%     3
684615.03%     3
              ..
245357.08%     1
507566.27%     1
321497.94%     1
1227534.31%    1
899704.02%     1
Name: count, Length: 1027, dtype: int64


Conclusions for 'Number of Open Complaints':
- Number of unique values: 6
- Most frequent value (top): 1/0/00 (occurs 830 times)
- Frequency distribution:
Number of Open Complaints
1/0/00    830
1/1/00    138
1/2/00     50
1/3/00     34
1/4/00     13
1/5/00      6
Name: count, dtype: int64


Conclusions for 'Policy Type':
- Number of unique values: 3
- Most frequent value (top): Personal Auto (occurs 780 times)
- Frequency distribution:
Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64


Conclusions for 'Vehicle Class':
- Number of unique values: 6
- Most frequent value (top): Four-Door Car (occurs 576 times)
- Frequency distribution:
Vehicle Class
Four-Door Car    576
Two-Door Car     205
SUV              199
Sports Car        57
Luxury SUV        20
Luxury Car        14
Name: count, dtype: int64


## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [ ]:
# Your code here

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [ ]:
# Your code here

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [ ]:
# Your code here

### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data.

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [ ]:
# Your code here